File: frontend/src/lib/seo-slug.ts
/**
 * Mirror of `shopifyseo.seo_slug.slugify_article_handle` for preview / default URL slug in the UI.
 */
export function slugifyArticleHandle(text: string, maxLen = 96): string {
  const raw = (text || "").trim();
  if (!raw) {
    return "article";
  }

  const normalized = raw.normalize("NFKD");
  let ascii = "";
  for (let i = 0; i < normalized.length; i++) {
    const code = normalized.charCodeAt(i);
    if (code <= 0x7f) {
      ascii += normalized[i];
    }
  }

  let slug = ascii.toLowerCase().replace(/[^a-z0-9]+/g, "-");
  slug = slug.replace(/-+/g, "-").replace(/^-|-$/g, "");

  if (!slug) {
    return "article";
  }

  if (slug.length > maxLen) {
    slug = slug.slice(0, maxLen).replace(/-+$/, "").replace(/-+/g, "-").replace(/^-|-$/g, "");
  }

  return slug || "article";
}

/** Words that add no SEO value in a URL slug. */
const STOP_WORDS = new Set([
  "a","an","the","and","or","but","in","on","at","to","for","of","with",
  "is","are","was","were","be","been","being","have","has","had",
  "do","does","did","will","would","shall","should","may","might",
  "can","could","about","from","into","through","during","before",
  "after","above","below","between","under","over","out","up","down",
  "off","then","than","so","no","not","only","very","just","how",
  "what","when","where","which","who","whom","why","all","each",
  "every","both","few","more","most","other","some","such","own",
  "same","too","also","your","you","its","our","their","my","this",
  "that","these","those","here","there","again","once","i","we","he",
  "she","it","they","me","him","her","us","them","need","know",
  "everything","nothing","something","anything","guide","complete",
  "ultimate","best","top","rated",
]);

/**
 * Build a concise, keyword-rich slug source from topic + keywords.
 *
 * Strategy:
 * 1. Extract meaningful words from the topic (strip stop words + filler).
 * 2. Prepend any keywords that aren't already present in the topic words.
 * 3. Target 3-5 words total — enough for search intent, short enough for clean URLs.
 *
 * Example:
 *   topic = "Everything You Need to Know About SMOK Novo Pod Systems"
 *   keywords = "novo, pod, coils"
 *   → "smok-novo-pod-systems"
 */
export function buildDraftSlugSource(topic: string, keywordsCsv: string): string {
  const kw = keywordsCsv
    .split(",")
    .map((k) => k.trim().toLowerCase())
    .filter(Boolean);

  // Extract meaningful words from the topic
  const topicWords = topic
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, " ")
    .split(/\s+/)
    .filter((w) => w.length >= 2 && !STOP_WORDS.has(w));

  // Start with topic words (up to 5), then add unique keyword terms
  const seen = new Set<string>();
  const parts: string[] = [];

  for (const w of topicWords) {
    if (!seen.has(w) && parts.length < 5) {
      seen.add(w);
      parts.push(w);
    }
  }

  // Add keyword terms not already covered by the topic
  for (const phrase of kw) {
    for (const w of phrase.split(/\s+/)) {
      const clean = w.replace(/[^a-z0-9]/g, "");
      if (clean.length >= 2 && !seen.has(clean) && parts.length < 5) {
        seen.add(clean);
        parts.push(clean);
      }
    }
  }

  return parts.length > 0 ? parts.join(" ") : topic.trim();
}

export function defaultDraftSlugHint(topic: string, keywordsCsv: string): string {
  return slugifyArticleHandle(buildDraftSlugSource(topic, keywordsCsv));
}

File: shopifyseo/seo_slug.py
"""URL-safe handles for blog articles (Shopify `handle` field)."""

from __future__ import annotations

import re
import unicodedata


def slugify_article_handle(text: str, *, max_len: int = 96) -> str:
    """Turn a title or topic into a Shopify-style handle: lowercase, hyphens, a-z0-9 only.

    Strips accents, collapses punctuation to single hyphens, trims length for readable URLs.
    """
    raw = (text or "").strip()
    if not raw:
        return "article"

    normalized = unicodedata.normalize("NFKD", raw)
    ascii_only = normalized.encode("ascii", "ignore").decode("ascii")
    lower = ascii_only.lower()
    slug = re.sub(r"[^a-z0-9]+", "-", lower)
    slug = re.sub(r"-{2,}", "-", slug).strip("-")

    if not slug:
        return "article"

    if len(slug) > max_len:
        slug = slug[:max_len].rstrip("-")

    return slug or "article"


# Words that add no ranking value in a URL slug.
_SLUG_STOP_WORDS: set[str] = {
    "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
    "of", "with", "is", "are", "was", "were", "be", "been", "being",
    "have", "has", "had", "do", "does", "did", "will", "would", "shall",
    "should", "may", "might", "can", "could", "about", "from", "into",
    "through", "during", "before", "after", "above", "below", "between",
    "under", "over", "out", "up", "down", "off", "then", "than", "so",
    "no", "not", "only", "very", "just", "how", "what", "when", "where",
    "which", "who", "whom", "why", "all", "each", "every", "both", "few",
    "more", "most", "other", "some", "such", "own", "same", "too", "also",
    "your", "you", "its", "our", "their", "my", "this", "that", "these",
    "those", "here", "there", "again", "once", "i", "we", "he", "she",
    "it", "they", "me", "him", "her", "us", "them", "need", "know",
    "everything", "nothing", "something", "anything", "guide", "complete",
    "ultimate", "best", "top", "rated",
}


def seo_article_slug(
    title: str,
    *,
    keywords: list[str] | None = None,
    max_words: int = 5,
    max_len: int = 60,
) -> str:
    """Build a concise, keyword-rich URL slug from the article title and optional keywords.

    Strategy:
    1. Extract meaningful words from the title (strip stop words + filler).
    2. Append any keyword terms not already present.
    3. Target *max_words* words — enough for search intent, short enough for clean URLs.

    Examples:
        title="Everything You Need to Know About SMOK Novo Pod Systems"
        keywords=["novo", "pod", "coils"]
        → "smok-novo-pod-systems-coils"
    """
    raw = (title or "").strip()
    if not raw:
        return slugify_article_handle("article")

    # Normalise to ASCII lowercase words
    normalized = unicodedata.normalize("NFKD", raw)
    ascii_only = normalized.encode("ascii", "ignore").decode("ascii").lower()
    words = re.sub(r"[^a-z0-9\s-]", " ", ascii_only).split()
    meaningful = [w for w in words if len(w) >= 2 and w not in _SLUG_STOP_WORDS]

    seen: set[str] = set()
    parts: list[str] = []
    for w in meaningful:
        if w not in seen and len(parts) < max_words:
            seen.add(w)
            parts.append(w)

    # Fill remaining slots with keyword terms not already in the title
    for kw_phrase in keywords or []:
        for w in re.sub(r"[^a-z0-9\s]", " ", kw_phrase.lower()).split():
            if len(w) >= 2 and w not in seen and len(parts) < max_words:
                seen.add(w)
                parts.append(w)

    if not parts:
        return slugify_article_handle(raw, max_len=max_len)

    slug = "-".join(parts)
    if len(slug) > max_len:
        slug = slug[:max_len].rstrip("-")
        if "-" in slug:
            slug = slug.rsplit("-", 1)[0]

    return slug or slugify_article_handle(raw, max_len=max_len)
